### PR TITLE
Build and test custom images locally

### PR DESCRIPTION
**Pull request by hurtadosanti**
https://hivemq.kanbanize.com/ctrl_board/22/cards/6566/details/

there is at the moment a dependency that is not fully updated on the hivemq-operator, so can not be push to master until is finished

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -0,0 +1,6 @@
+# Changelog
+## hivemq-operator-0.10.9
+### Add integration tests
+- Custom images tests
+- Matrix tests
+- Security rootless images tests
```
Comment (YannickWeber) [inline CHANGELOG.md:6]: Please check other projects to copy the format for the CHANGELOG.md file

File: src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+
+    @Timeout(value = 20, unit = TimeUnit.MINUTES)
+    @ParameterizedTest
+    @ValueSource(strings = {"v1.20.15-k3s1", "v1.21.10-k3s1", "v1.22.7-k3s1", "v1.23.4-k3s1"})
+    public void withHelmVersionDeployment_mqttMessagePublishedReceived(final @NotNull String version) throws Exception {
+        try (var container = new OperatorHelmChartContainer(version, "k3s.dockerfile",
+                "values/customTestValues.yaml")) {
+            container.withCustomImages();
+            container.start();
+            Mqtt5BlockingClient client = Mqtt5Client.builder()
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java:35]:
```suggestion
            final Mqtt5BlockingClient client = Mqtt5Client.builder()
```

File: src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+ * Test that the chart is deployed successfully on specific kubernetes cluster versions
+ */
+@Testcontainers
+public class CompatibilityHelmChartDeploymentIT {
+
+    @Timeout(value = 20, unit = TimeUnit.MINUTES)
+    @ParameterizedTest
+    @ValueSource(strings = {"v1.20.15-k3s1", "v1.21.10-k3s1", "v1.22.7-k3s1", "v1.23.4-k3s1"})
+    public void withHelmVersionDeployment_mqttMessagePublishedReceived(final @NotNull String version) throws Exception {
+        try (var container = new OperatorHelmChartContainer(version, "k3s.dockerfile",
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java:31]:
```suggestion
        try (final var container = new OperatorHelmChartContainer(version, "k3s.dockerfile",
```

File: src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+            container.withCustomImages();
+            container.start();
+            Mqtt5BlockingClient client = Mqtt5Client.builder()
+                    .automaticReconnectWithDefaultConfig()
+                    .serverPort(container.getMappedPort(1883))
+                    .buildBlocking();
+
+            client.connect();
+
+            var publishes = client.publishes(MqttGlobalPublishFilter.ALL);
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java:42]:
```suggestion
           final var publishes = client.publishes(MqttGlobalPublishFilter.ALL);
```

File: src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+            client.subscribeWith().topicFilter("test").send();
+            client.publishWith()
+                    .topic("test")
+                    .payload("Sending Message".getBytes(StandardCharsets.UTF_8))
+                    .qos(MqttQos.AT_LEAST_ONCE).send();
+
+            Mqtt5Publish receivedMessage = publishes.receive();
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java:50]:
```suggestion
           final Mqtt5Publish receivedMessage = publishes.receive();
```

File: src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java [hunk 6]
```diff
@@ -0,0 +1,60 @@
...
+    @ParameterizedTest
+    @ValueSource(strings = {"v1.20.15-k3s1", "v1.21.10-k3s1", "v1.22.7-k3s1", "v1.23.4-k3s1"})
+    public void withHelmVersionDeployment_mqttMessagePublishedReceived(final @NotNull String version) throws Exception {
+        try (var container = new OperatorHelmChartContainer(version, "k3s.dockerfile",
+                "values/customTestValues.yaml")) {
+            container.withCustomImages();
+            container.start();
+            Mqtt5BlockingClient client = Mqtt5Client.builder()
+                    .automaticReconnectWithDefaultConfig()
+                    .serverPort(container.getMappedPort(1883))
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java:37]:
```suggestion
                    .serverPort(container.getMappedPort(1883))
                    .serverHost(container.getHost())
```

File: src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java [hunk 7]
```diff
@@ -0,0 +1,60 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+/**
+ * Test that the chart is deployed successfully on specific kubernetes cluster versions
+ */
+@Testcontainers
+public class CompatibilityHelmChartDeploymentIT {
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CompatibilityHelmChartDeploymentIT.java:25]: Please be more consistent when using var or explicit type declarations. Either one or the other.

File: src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+    private static final @NotNull OperatorHelmChartContainer
+            container = new OperatorHelmChartContainer("v1.23.4-k3s1",
+            "k3s.dockerfile",
+            "values/customTestValues.yaml")
+            .withCustomImages();
+
+    @Timeout(value = 3, unit = TimeUnit.MINUTES)
+    @Test
+    public void withCustomImage_mqttMessagePublishedReceived() throws Exception {
+        assertTrue(container.isRunning());
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java:37]: remove

File: src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+        //System.out.println(Runtime.getRuntime().maxMemory());
+        Mqtt5BlockingClient client = Mqtt5Client.builder()
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java:39]: final

File: src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                .automaticReconnectWithDefaultConfig()
+                .serverPort(container.getMappedPort(1883))
+                .buildBlocking();
+
+        client.connect();
+
+        var publishes = client.publishes(MqttGlobalPublishFilter.ALL);
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java:46]: final

File: src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+        client.subscribeWith().topicFilter("test").send();
+        client.publishWith()
+                .topic("test")
+                .payload("Sending Message".getBytes(StandardCharsets.UTF_8))
+                .qos(MqttQos.AT_LEAST_ONCE).send();
+
+        Mqtt5Publish receivedMessage = publishes.receive();
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java:54]: final

File: src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+        assertTrue(receivedMessage.getPayload().isPresent());
+        assertEquals("Sending Message", StandardCharsets.UTF_8.decode(receivedMessage
+                .getPayload().get().asReadOnlyBuffer()).toString());
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomHelmChartDeploymentIT.java:57]:
```suggestion
        assertEquals("Sending Message", StandardCharsets.UTF_8.decode(receivedMessage
                .getPayload().get().toString());
```

File: src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    private static final @NotNull K3sContainer container = new K3sContainer(DockerImageName
+            .parse("rancher/k3s:v1.21.3-k3s1"))
+            .withFileSystemBind("./build/containers", "/build", BindMode.READ_ONLY)
+            .withFileSystemBind("./charts/hivemq-operator", "/chart");
+
+    @Timeout(value = 2, unit = TimeUnit.MINUTES)
+    @Test
+    public void withCustomImage_mqttMessagePublishedReceived() throws Exception {
+
+        System.out.println(Runtime.getRuntime().maxMemory());
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:30]: remove

File: src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.jupiter.api.Assertions.assertFalse;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@Testcontainers
+public class CustomImageDeploymentIT {
+
+    @Container
+    private static final @NotNull K3sContainer container = new K3sContainer(DockerImageName
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:21]: why is this static?

File: src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+            .withFileSystemBind("./build/containers", "/build", BindMode.READ_ONLY)
+            .withFileSystemBind("./charts/hivemq-operator", "/chart");
+
+    @Timeout(value = 2, unit = TimeUnit.MINUTES)
+    @Test
+    public void withCustomImage_mqttMessagePublishedReceived() throws Exception {
+
+        System.out.println(Runtime.getRuntime().maxMemory());
+
+        var containerName = "hivemq-k8s-image.tar";
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:32]: inline or create a constant

File: src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java [hunk 16]
```diff
@@ -0,0 +1,45 @@
...
+
+        var outLoadImage = container.execInContainer("/bin/ctr",
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:34]: final

File: src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+                "images",
+                "import",
+                "/build/" + containerName);
+
+        assertFalse(outLoadImage.getStdout().isEmpty());
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:39]: I think you can assert the exit code of that command.

Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:39]: Please add a comment, why you are testing that the stdout is not empty.
you can also assert that the exit code is 0

File: src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java [hunk 18]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.concurrent.TimeUnit;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@Testcontainers
+public class UserPermissionsIT {
+
+    @Container
+    private static final @NotNull OperatorHelmChartContainer
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java:25]:
```suggestion
    private final @NotNull OperatorHelmChartContainer
```

File: src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+            container = new OperatorHelmChartContainer("v1.23.4-k3s1",
+            "k3s.dockerfile",
+            "values/permissionsDeployment.yaml")
+            .withCustomImages("hivemq-k8s-test-rootless.tar");
+
+    @Timeout(value = 4, unit = TimeUnit.MINUTES)
+    @Test
+    public void withCustomImage_mqttMessagePublishedReceived() throws Exception {
+        assertTrue(container.isRunning());
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java:34]:
```suggestion

```

File: src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+        //System.out.println(Runtime.getRuntime().maxMemory());
+        Mqtt5BlockingClient client = Mqtt5Client.builder()
+                .automaticReconnectWithDefaultConfig()
+                .serverPort(container.getMappedPort(1883))
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java:38]: also use container host

File: src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+                .buildBlocking();
+
+        client.connect();
+
+        var publishes = client.publishes(MqttGlobalPublishFilter.ALL);
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java:43]: final

File: src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+
+        client.subscribeWith().topicFilter("test").send();
+        client.publishWith()
+                .topic("test")
+                .payload("Sending Message".getBytes(StandardCharsets.UTF_8))
+                .qos(MqttQos.AT_LEAST_ONCE).send();
+
+        Mqtt5Publish receivedMessage = publishes.receive();
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/UserPermissionsIT.java:51]: final

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 23]
```diff
@@ -1,85 +1,164 @@
...
+import java.util.concurrent.CountDownLatch;
+
+import static org.junit.jupiter.api.Assertions.*;
+import static org.testcontainers.containers.output.OutputFrame.OutputType.STDERR;
 
 /**
  * Container that includes the helm binary to be able to install the HiveMQ helm charts
  */
 public class OperatorHelmChartContainer extends K3sContainer {
     public final static int mqttPort = 1883;
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:34]: constants are Upper case

File: build.gradle.kts [hunk 24]
```diff
@@ -15,8 +15,13 @@ repositories {
         url = uri("https://jitpack.io")
     }
 }
+val hivemq: Configuration by configurations.creating { isCanBeConsumed = false; isCanBeResolved = false }
+
+val operator: Configuration by configurations.creating { isCanBeConsumed = false; isCanBeResolved = false }
 
 dependencies {
+    hivemq("com.hivemq:hivemq")
+    operator("com.hivemq:hivemq-operator")
```
Comment (YannickWeber) [inline build.gradle.kts:24]: Can't this be achieved with a https://docs.gradle.org/current/userguide/composite_builds.html ?

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 25]
```diff
@@ -1,85 +1,164 @@
...
 
-        public @NotNull Builder helmChartMountPath(final @NotNull File mountPath) {
-            this.helmChartMountPath = mountPath;
-            return this;
+        @Override
+        public @NotNull StartupStatus checkStartupState(@NotNull DockerClient dockerClient, @NotNull String containerId) {
+            var s = container.getLogs(STDERR);
+            while (!s.matches("(?s).*Node controller sync successful.*")) {
+                s = container.getLogs(STDERR);
+            }
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:88]: Replace this spinning log with https://mvnrepository.com/artifact/org.awaitility/awaitility or use https://github.com/hivemq/hivemq-security-extension/blob/6415475e151c2a7818768a8b75b3716838261a47/src/integrationTest/java/com/hivemq/extensions/security/util/LogWaiterUtil.java

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 26]
```diff
@@ -1,85 +1,164 @@
...
+            try {
+                // we need this to have the yaml read from the container
+                container.containerIsStarted(container.getContainerInfo());
+                var yaml = container.getKubeConfigYaml();
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:92]: final

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 27]
```diff
@@ -1,85 +1,164 @@
...
-            return this;
+        public void loadImages() {
+            container.getImagesNames().forEach(a -> {
+                try {
+                    var outLoadImage = container.execInContainer("/bin/ctr",
+                            "images",
+                            "import",
+                            "/build/" + a);
+                    assertFalse(outLoadImage.getStdout().isEmpty());
+                } catch (IOException | InterruptedException e) {
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:113]:
```suggestion
                } catch (final Exception e) {
```

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 28]
```diff
@@ -1,85 +1,164 @@
...
         }
 
-        public @NotNull Builder containerPath(final @NotNull String containerPath) {
-            this.containerPath = containerPath;
-            return this;
+        @SuppressWarnings("resource")
+        private void waitForClusterToBeReady(final @NotNull String kubeConfigYaml) throws InterruptedException {
+            final CountDownLatch closeLatch = new CountDownLatch(1);
+            Config config = Config.fromKubeconfig(kubeConfigYaml);
+            DefaultKubernetesClient client = new DefaultKubernetesClient(config);
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:124]: final

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 29]
```diff
@@ -1,85 +1,164 @@
...
+            client.customResources(HiveMQInfo.class).watch(new Watcher<>() {
+                @Override
+                public void eventReceived(@NotNull Action action, @NotNull HiveMQInfo resource) {
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:127]:
```suggestion
                public void eventReceived(final @NotNull Action action, final @NotNull HiveMQInfo resource) {
```

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 30]
```diff
@@ -1,85 +1,164 @@
...
+
+                    if (resource.getStatus() != null
+                            && resource.getStatus().getState() != null
+                            && resource.getStatus().getState() == HivemqClusterStatus.State.RUNNING) {
+                        closeLatch.countDown();
+                    }
+                }
+
+                @Override
+                public void onClose(@NotNull WatcherException cause) {
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:137]:
```suggestion
                public void onClose(final @NotNull WatcherException cause) {
```

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 31]
```diff
@@ -1,85 +1,164 @@
...
-                                       final @NotNull String containerPath) {
-        super(getAdHocImageName(k3sVersion, dockerfile));
-        final MountableFile helmChartMountable = MountableFile.forHostPath(helmChartMountPath.getPath());
-        super.withCopyFileToContainer(helmChartMountable, containerPath);
+    public OperatorHelmChartContainer(final @NotNull String k3sVersion,
+                                      final @NotNull String dockerfileName,
+                                      final @NotNull String customValuesFile) {
+        super(getAdHocImageName(k3sVersion, dockerfileName));
         super.addExposedPort(mqttPort);
+        super.withFileSystemBind("./charts/hivemq-operator", "/chart");
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:43]: Use `withCopyFileToContainer` instead of `withFileSystemBind`. `withCopyFileToContainer` seems to be more stable on CIs.

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 32]
```diff
@@ -1,85 +1,164 @@
...
     }
 
-    public int getMappedPort() {
-        return super.getMappedPort(mqttPort);
+    /**
+     * Uses custom images instead of docker hub images, additional images can be added
+     */
+    public @NotNull OperatorHelmChartContainer withCustomImages(@Nullable String... fileNames) {
+        withCustomImages = true;
+        super.withFileSystemBind("./build/containers", "/build", BindMode.READ_ONLY);
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:65]: Use `withCopyFileToContainer` instead of `withFileSystemBind`. `withCopyFileToContainer` seems to be more stable on CIs.

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 33]
```diff
@@ -1,85 +1,164 @@
...
-        return new Builder();
+        return DockerImageName.parse(s).asCompatibleSubstituteFor("rancher/k3s");
     }
 
-    public int getMappedPort() {
-        return super.getMappedPort(mqttPort);
+    /**
+     * Uses custom images instead of docker hub images, additional images can be added
+     */
+    public @NotNull OperatorHelmChartContainer withCustomImages(@Nullable String... fileNames) {
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:63]:
```suggestion
    public @NotNull OperatorHelmChartContainer withCustomImages(final @Nullable String... fileNames) {
```

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 34]
```diff
@@ -1,85 +1,164 @@
...
+        imagesNames = new ArrayList<>();
     }
 
-    public static @NotNull DockerImageName getAdHocImageName(final @NotNull String k3sVersion,
-                                                             final @NotNull File dockerfile) {
-        final String s = new ImageFromDockerfile().withDockerfile(
-                        dockerfile.toPath())
+    private static @NotNull DockerImageName getAdHocImageName(final @NotNull String k3sVersion,
+                                                             final @NotNull String dockerfileName) {
+        var dockerfile = new File(MountableFile.forClasspathResource(dockerfileName).getFilesystemPath());
```
Comment (YannickWeber) [inline src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:51]:
```suggestion
        final var dockerfile = new File(MountableFile.forClasspathResource(dockerfileName).getFilesystemPath());
```

**Review by YannickWeber — CHANGES_REQUESTED** [review]
Looks good so far, I have some comments.
Please make sure that the "fieldMayBeFinal" inspection is activated in your IDE.

File: src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java [hunk 35]
```diff
@@ -0,0 +1,45 @@
...
+import static org.junit.jupiter.api.Assertions.assertTrue;
+
+@Testcontainers
+public class CustomImageDeploymentIT {
+
+    @Container
+    private static final @NotNull K3sContainer container = new K3sContainer(DockerImageName
+            .parse("rancher/k3s:v1.21.3-k3s1"))
+            .withFileSystemBind("./build/containers", "/build", BindMode.READ_ONLY)
+            .withFileSystemBind("./charts/hivemq-operator", "/chart");
```
[the remark this answers is not in the document]
Comment (hurtadosanti) [reply src/integrationTest/java/com/hivemq/helmcharts/CustomImageDeploymentIT.java:24]: I have issues with big files, that is why I am using the System Bind

File: src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java [hunk 36]
```diff
@@ -1,85 +1,164 @@
...
+            });
+
         }
 
-        public @NotNull Builder containerPath(final @NotNull String containerPath) {
-            this.containerPath = containerPath;
-            return this;
+        @SuppressWarnings("resource")
+        private void waitForClusterToBeReady(final @NotNull String kubeConfigYaml) throws InterruptedException {
+            final CountDownLatch closeLatch = new CountDownLatch(1);
```
[the remark this answers is not in the document]
Comment (hurtadosanti) [reply src/integrationTest/java/com/hivemq/helmcharts/util/OperatorHelmChartContainer.java:122]: This changes was discuss with @Florian-Limpoeck,

File: build.gradle.kts (hunk 24 above)
Comment (hurtadosanti) [reply build.gradle.kts:24]: @SgtSilvio can you confirm if this is correct please?

File: charts/hivemq-operator/Chart.lock [hunk 37]
```diff
@@ -1,6 +0,0 @@
-dependencies:
```
Comment (SgtSilvio) [inline charts/hivemq-operator/Chart.lock:1]: is this file not necessary anymore?

File: build.gradle.kts (hunk 24 above)
Comment (SgtSilvio) [reply build.gradle.kts:24]: These are just dependencies, and actually included builds are used.

File: build.gradle.kts [hunk 38]
```diff
@@ -33,8 +38,16 @@ dependencies {
     testImplementation("org.awaitility:awaitility:${property("awaitility.version")}")
 }
 
-tasks.getByName<Test>("test") {
+tasks.withType<Test>().configureEach {
     useJUnitPlatform()
+    maxHeapSize="6g"
```
Comment (SgtSilvio) [inline build.gradle.kts:43]: just some formatting nitpicks
```suggestion
    maxHeapSize = "6g"
```

File: build.gradle.kts [hunk 39]
```diff
@@ -47,4 +60,72 @@ val integrationTestImplementation: Configuration by configurations.getting {
...
+    attributes {
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named("operator-docker-image"))
+    }
+    extendsFrom(operator)
+}
+
+
+val createRootlessK8sImageContext by tasks.registering(Sync::class) {
+    group = "container"
+    description = "Prepare container base image context"
```
Comment (SgtSilvio) [inline build.gradle.kts:95]:
```suggestion
    description = "Prepare rootless k8s image context"
```

File: build.gradle.kts [hunk 40]
```diff
@@ -47,4 +60,72 @@ val integrationTestImplementation: Configuration by configurations.getting {
...
+    into(layout.buildDirectory.dir("container/context"))
+    from("container")
+}
+
+val buildRootlessK8sImage by tasks.registering(Exec::class){
+    group = "container"
+    description = "Build docker image"
```
Comment (SgtSilvio) [inline build.gradle.kts:102]:
```suggestion
    description = "Build rootless k8s image"
```

File: build.gradle.kts [hunk 41]
```diff
@@ -47,4 +60,72 @@ val integrationTestImplementation: Configuration by configurations.getting {
...
+        attribute(Category.CATEGORY_ATTRIBUTE, objects.named("operator-docker-image"))
+    }
+    extendsFrom(operator)
+}
+
+
+val createRootlessK8sImageContext by tasks.registering(Sync::class) {
+    group = "container"
+    description = "Prepare container base image context"
+    into(layout.buildDirectory.dir("container/context"))
```
Comment (SgtSilvio) [inline build.gradle.kts:96]:
```suggestion
    into(layout.buildDirectory.dir("container/context/hivemq-rootless-k8s-image"))
```

File: build.gradle.kts [hunk 42]
```diff
@@ -47,4 +60,72 @@ val integrationTestImplementation: Configuration by configurations.getting {
...
+val buildRootlessK8sImage by tasks.registering(Exec::class){
+    group = "container"
+    description = "Build docker image"
+    inputs.property("dockerImageName", containerName)
+    inputs.dir(createRootlessK8sImageContext.map { it.destinationDir })
+    workingDir(createRootlessK8sImageContext.map { it.destinationDir })
+    commandLine("docker","build","-f","broker.dockerfile","-t","${containerName}-rootless:${containerTag}",".")
+}
+
+val saveRootlessK8sImage by tasks.registering(Exec::class){
```
Comment (SgtSilvio) [inline build.gradle.kts:109]: As you have 2 tasks for building and saving the docker image, the tasks can never be up-to-date as inputs and outputs are needed. This means the tasks are always re-executed, which will always take a lot of time (especially for the `saveRootlessK8sImage` task).

File: build.gradle.kts [hunk 43]
```diff
@@ -47,4 +60,72 @@ val integrationTestImplementation: Configuration by configurations.getting {
...
+    dependsOn(buildRootlessK8sImage)
+    workingDir(layout.buildDirectory.dir("containers"))
+    commandLine("docker","save","-o","${containerName}-rootless.tar","${containerName}-rootless:${containerTag}")
+}
+val buildContainersFiles by tasks.registering(Copy::class){
+    group = "container"
+    dependsOn(gradle.includedBuild("hivemq").task(":buildK8sImage"))
+    dependsOn(gradle.includedBuild("hivemq-operator").task(":saveDnsInitWaitImage"))
+    dependsOn(gradle.includedBuild("hivemq-operator").task(":jibBuildTar"))
+    from(producerK8sDockerImage.singleFile,producerDnsInitWaitDockerImage.singleFile,producerOperatorDockerImage.singleFile)
```
Comment (SgtSilvio) [inline build.gradle.kts:121]: This should be enough, dependencies are automatically added
```suggestion
    from(producerK8sDockerImage)
    from(producerDnsInitWaitDockerImage)
    from(producerOperatorDockerImage)
```

File: build.gradle.kts [hunk 44]
```diff
@@ -47,4 +60,72 @@ val integrationTestImplementation: Configuration by configurations.getting {
...
+    into(layout.buildDirectory.dir("containers"))
+}
+
+buildContainersFiles{
+    finalizedBy(saveRootlessK8sImage)
```
[the remark this answers is not in the document]
Comment (hurtadosanti) [reply build.gradle.kts:126]: Yes, the order matter here, I know is not nice but otherwise it will not finish

File: build.gradle.kts (hunk 43 above)
Comment (hurtadosanti) [reply build.gradle.kts:121]: This I need 'dependsOn(gradle.includedBuild("hivemq-operator").task(":jibBuildTar"))' to be run

File: build.gradle.kts (hunk 42 above)
Comment (hurtadosanti) [reply build.gradle.kts]: I leave it for next fix as will include changes to the images

File: charts/hivemq-operator/Chart.lock (hunk 37 above)
Comment (hurtadosanti) [reply charts/hivemq-operator/Chart.lock:1]: Added back
